components/ai_chat/ai_chat_api.cc

```diff
@@ -78,27 +82,63 @@ AIChatAPI::AIChatAPI(
     CHECK(api_base_url.is_valid()) << "API Url generated was invalid."
                                       "Please check configuration parameter.";
   }
+
+  // We need our own instance of DataDecoder and remote for JsonParser to ensure
+  // ordered processing of stream chunks. Calling the static function creates a
+  // new instance of the process for each call, which may result in unordered
+  // chunks. Thus, we create a single instance of the parser and reuse it for
+  // consecutive calls
+  data_decoder_ = std::make_unique<data_decoder::DataDecoder>();
+  data_decoder_->GetService()->BindJsonParser(
+      json_parser_.BindNewPipeAndPassReceiver());
 }
 
 AIChatAPI::~AIChatAPI() = default;
 
-void AIChatAPI::QueryPrompt(ResponseCallback callback,
+bool AIChatAPI::IsRequestInProgress() {
+  return is_request_in_progress_;
+}
+
+void AIChatAPI::QueryPrompt(ResponseCallback response_callback,
+                            CompletionCallback completion_callback,
                             const std::string& prompt) {
   const GURL api_base_url = GetEndpointBaseUrl();
-  // Verify that we have a Url configured
-  if (api_base_url.is_empty()) {
-    std::move(callback).Run("", false);
-    return;
-  }
+
   // Validate that the path is valid
   GURL api_url = api_base_url.Resolve(ai_chat::kAIChatCompletionPath);
   CHECK(api_url.is_valid())
       << "Invalid API Url, check path: " << api_url.spec();
 
-  auto internal_callback =
-      base::BindOnce(&AIChatAPI::OnGetResponse, weak_ptr_factory_.GetWeakPtr(),
-                     std::move(callback));
+  const base::Value::Dict& dict = CreateApiParametersDict(prompt);
+  base::flat_map<std::string, std::string> headers;
+  headers.emplace("x-brave-key", BUILDFLAG(BRAVE_SERVICES_KEY));
+  headers.emplace("Accept", "text/event-stream");
+
+  if (IsRequestInProgress()) {
+    // TODO(nullhook): Send an event for this so UI can reject further inputs
+    DVLOG(1) << __func__ << "There is a request in progress\n";
+    return;
+  }
 
+  // TODO(@nullhook): Add a response started callback
+  response_callback_ = response_callback;
+  completion_callback_ = std::move(completion_callback);
```

Comment:
Storing the callbacks in a single variable makes AIChatAPI definitely something that can only handle one request at a time, and I'm not sure that's clear or if it's neccessary. Can references to them not be passed to the response handler?

Reply:
This has been addressed  in the new architecture